scripts/migrate_data

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+"""
+Migrate .fits files whose format has changed between jwst package versions.
+"""
+import argparse
+from datetime import datetime
+import os
+import re
+import traceback
+import warnings
+
+import asdf
+from astropy.io import fits
+import numpy as np
+from packaging.specifiers import SpecifierSet
+
+import jwst
+from jwst import datamodels
+
+
+def parse_args():
+    parser = argparse.ArgumentParser("migrate_data", "migrate .fits files whose format has changed between jwst package versions")
+
+    parser.add_argument("files", nargs="+", help="one or more .fits files")
+
+    output_group = parser.add_mutually_exclusive_group(required=True)
+    output_group.add_argument("--output-dir", help="write modified files to an output directory")
+    output_group.add_argument("--in-place", help="modify files in-place", action="store_true")
+
+    return parser.parse_args()
+
+
+# If there get to be many of these we may want to move
+# them to jwst.datamodels somewhere:
+
+def rebuild_spec_table(table):
+    """
+    spectable.schema added additional columns and renamed
+    two columns.
+    """
+    schema = asdf.schema.load_schema("http://stsci.edu/schemas/jwst_datamodel/spectable.schema")
+    dtype = asdf.tags.core.ndarray.asdf_datatype_to_numpy_dtype(schema["datatype"])
+    renamed_columns = {
+        "ERROR": "FLUX_ERROR",
+        "BERROR": "BKGD_ERROR",
+    }
+
+    new_table = np.zeros(table.shape, dtype=dtype)
+    for column_name in table.dtype.names:
+        new_table[renamed_columns.get(column_name, column_name)] = table[column_name]
+
+    return new_table
+
+
+def migrate_spec_model_1_1_0(hdul):
+    for hdu in hdul:
+        if hdu.name == "EXTRACT1D":
+            hdu.data = rebuild_spec_table(hdu.data)
+
+
+def migrate_multispec_model_1_1_0(hdul):
+    for hdu in hdul:
+        if hdu.name == "EXTRACT1D":
+            hdu.data = rebuild_spec_table(hdu.data)
+
+
+# The first key is a model class name, the second
+# a jwst package version specifier.  The value
+# is a method that accepts an HDUList and modifies
+# it in-place.
+_MIGRATE_METHODS = {
+    "SpecModel": {
+        "<= 1.1.0": migrate_spec_model_1_1_0,
+    },
+    "MultiSpecModel": {
+        "<= 1.1.0": migrate_multispec_model_1_1_0,
+    },
+}
+
+
+def migrate_file(filename, args):
+    if args.in_place:
+        mode = "update"
+    else:
+        mode = "readonly"
+
+    with fits.open(filename, memmap=False, mode=mode) as hdul:
+        model_type = hdul[0].header.get("DATAMODL")
+        jwst_version = hdul[0].header.get("CAL_VER")
+
+        if not (model_type and jwst_version):
+            print(f"Unable to migrate {filename}: DATAMODL and CAL_VER keywords are required")
+            return
+
+        match = re.match(r'^[0-9]+\.[0-9]+\.[0-9]+', jwst_version)
```

Comment:
So does this checking of CAL_VER value mean that the script will only work for x1d's that were created using 1.1.0? We might need to allow it to go back a bit further than that, whenever the `SpecModel` had it's last change before this big one.

Reply:
The requirement `"<= 1.1.0"` is interpreted as "any cal version that satisfies this requirement needs to be migrated".  So, it's going to try to migrate any version before 1.1.0 as well.  Maybe that's too many?  It looks like the change that removed NET and added SURF_BRIGHT was released in 0.13.2, so I can make that the lower bound.

Reply:
Sounds good.